pilot/pkg/config/clusterregistry/secretcontroller.go

```diff
@@ -0,0 +1,158 @@
+	secretsSynced cache.InformerSynced
+	cs            *ClusterStore
+}
+
+// NewController returns a new secret controller
+func NewController(
+	kubeclientset kubernetes.Interface,
+	kubeInformerFactory kubeinformers.SharedInformerFactory,
+	namespace string,
+	cs *ClusterStore) *Controller {
+
+	secretsInformer := kubeInformerFactory.Core().V1().Secrets()
+
+	controller := &Controller{
+		kubeclientset: kubeclientset,
+		namespace:     namespace,
+		secretsSynced: secretsInformer.Informer().HasSynced,
+		cs:            cs,
+	}
+
+	log.Info("Setting up event handlers")
+	secretsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.secretAdd,
+		DeleteFunc: controller.secretDelete,
+	})
+
+	return controller
+}
+
+// Run starts the controller until it receves a message over stopCh
+func (c *Controller) Run(stopCh <-chan struct{}) error {
+	defer runtime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	log.Info("Starting Secrets controller")
+
+	// Wait for the caches to be synced before starting workers
+	log.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.secretsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	<-stopCh
+	log.Info("Shutting down workers")
+
+	return nil
+}
+
+func checkSecret(s *corev1.Secret) bool {
+	lv, ok := s.Labels[mcLabel]
+	if !ok {
+		return false
+	}
+	if strings.ToLower(lv) != "true" {
+		return false
+	}
+	_, ok = s.ObjectMeta.Annotations[mcAnnotation]
+
+	return ok
+}
+
+func addMemberCluster(s *corev1.Secret, cs *ClusterStore) {
+	cs.storeLock.Lock()
+	defer cs.storeLock.Unlock()
+	// Check if there is already a cluster member with the specified
+	if _, ok := cs.clientConfigs[s.ObjectMeta.Annotations[mcAnnotation]]; !ok {
+		log.Infof("Adding new cluster member: %s", s.ObjectMeta.Annotations[mcAnnotation])
+		// cs.store[s.ObjectMeta.Annotations[mcAnnotation]] = s.Data[s.ObjectMeta.Annotations[mcAnnotation]]
+	}
+	log.Infof("Number of clusters in the cluster store: %d", len(cs.clientConfigs))
+}
+
+func deleteMemberCluster(s *corev1.Secret, cs *ClusterStore) {
+	cs.storeLock.Lock()
+	defer cs.storeLock.Unlock()
+	// Check if there is a cluster member with the specified name
+	if _, ok := cs.clientConfigs[s.ObjectMeta.Annotations[mcAnnotation]]; ok {
+		log.Infof("Deleting cluster member: %s", s.ObjectMeta.Annotations[mcAnnotation])
+		delete(cs.clientConfigs, s.ObjectMeta.Annotations[mcAnnotation])
+	}
+	log.Infof("Number of clusters in the cluster store: %d", len(cs.clientConfigs))
+}
+
+func (c *Controller) secretAdd(obj interface{}) {
+	s := obj.(*corev1.Secret)
+	if s.ObjectMeta.Namespace == c.namespace {
+		if checkSecret(s) {
+			addMemberCluster(s, c.cs)
+		}
+	}
+}
+
+func (c *Controller) secretDelete(obj interface{}) {
+	s := obj.(*corev1.Secret)
+	if s.ObjectMeta.Namespace == c.namespace {
+		if checkSecret(s) {
+			deleteMemberCluster(s, c.cs)
```

Comment:
TODO: delete all endpoints/services loaded from this cluster.

Reply:
Interesting idea, at this point not sure what could trigger it. controller invokes functions for add/delete/update of a specific object type. Would be great to discuss it.

Reply:
Just add a TODO - we don't need to implement it now.